lib/diamonds.ts

```diff
@@ -0,0 +1,137 @@
+import { Contract, constants } from 'ethers';
+
+// TODO: import IDiamond type
+
+export interface Facet {
+  target: string;
+  selectors: string[];
+}
+
+enum FacetCutAction {
+  Add,
+  Replace,
+  Remove,
+}
+
+export interface FacetCut extends Facet {
+  action: FacetCutAction;
+}
+
+export function getContractSignatures(contract: Contract) {
+  return Object.keys(contract.interface.functions);
+}
+
+export function getContractSelectors(contract: Contract) {
+  const signatures = getContractSignatures(contract);
+  const selectors = signatures.reduce((acc: string[], val: string) => {
+    acc.push(contract.interface.getSighash(val));
+    return acc;
+  }, []);
+
+  return selectors;
+}
+
+export function selectorExistsInFacet(selector: string, facets: Facet[]) {
+  for (const facet of facets) {
+    if (facet.selectors.includes(selector)) return true;
+  }
+
+  return false;
+}
+
+export function printFacetCuts(
+  target: string,
+  selectors: string[],
+  action: number = 0,
+) {
+  return {
+    target: target,
+    action: action,
+    selectors: selectors,
+  };
+}
+
+// adds unregistered selectors
+export async function addUnregisteredSelectors(
+  diamond: Contract,
+  contracts: Contract[],
+) {
+  const registeredFacets: Facet[] = await diamond.facets();
```

Comment:
Do we need any try catches?

Could we use `@solidstate/typechain-types`?

Reply:
> Do we need any try catches?

Possibly, did you have anything in mind? I was considered adding some form of validation that would throw if some conditions could result in `diamondCut` reverting. In the end I decided to use if-statements to check these conditions before adding a facetCut. Not sure if this is what you had in mind though.

```js
if (
  target !== diamond.address &&
  selector.length > 0 &&
  !selectorExistsInFacet(selector, registeredFacets)
) {
  facetCuts.push(
    printFacetCuts(facet.target, [selector], FacetCutAction.Add),
  );
}
```

Reply:
> Could we use `@solidstate/typechain-types`?

Definitely, I have this on my list of things to do.